### PR TITLE
Update JAR references to currently used versions

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
The MCC module's jars.txt has fallen out of date with the versions being included as reported by this TeamCity test failure:

https://teamcity.labkey.org/repository/download/bt3/1539820:id/BasicTest.tar.gz!/202109142153BasicTest%23testCredits.html

#### Changes
* Update reference from 2.9.5 to 2.11.3

### CODE REVIEW

**Comment by bbimber** [comment]
@labkey-jeckels There's a gradle-based way to declare/maintain dependencies now, right?

**Comment by labkey-susanh** [comment]
> @labkey-jeckels There's a gradle-based way to declare/maintain dependencies now, right?

Yes, this can be done in Gradle. Look for usages of `BuildUtils.addExternalDependency` in `build.gradle` files.

**Review by labkey-susanh — APPROVED** [review]
If updated to use `BuildUtils.addExternalDependency` then the `jars.txt` file can go away.

**Comment by labkey-jeckels** [comment]
I will merge this as-is as a quick test fix for now. @bbimber, if you want to migrate to the build.gradle approach in your copy of the repo we can pick it up via the usual merge process.

**Comment by bbimber** [comment]
@labkey-jeckels ok, sure

**Comment by labkey-adam** [comment]
@labkey-jeckels @bbimber Looks like the DRTs have been failing with an MCC credits discrepancy for the last month: https://teamcity.labkey.org/viewType.html?buildTypeId=bt3&branch_LabkeyTrunk=%3Cdefault%3E&tab=buildTypeStatusDiv

**WARNING: The following JAR file was found in your webapp but is not documented in the MCC Module jars.txt:**
jackson-dataformat-xml-2.11.3.jar

**WARNING: The following JAR file is documented in the MCC Module jars.txt but was not found in your webapp:**
jackson-dataformat-2.11.3.jar

**Comment by bbimber** [comment]
It's on my todo list to update my lab's modules to be using BuildUtils.addExternalDependency. I'll try to get that done today or tomorrow....

**Comment by bbimber** [comment]
Hey - I'm updating our modules with addExternalDependency, but I wonder if someone better at gradle could weigh in on 2 things:

1) Is it true that any instance where our module gradle files have 'external', but not 'implementation' need records on jars.txt and should be converted to addExternalDependnecy?

2) What's the syntax to convert a more complex entry like this one (my attempt at converting below)? Specifically, the 'exclude' stuff:

```
   external ("io.repseq:repseqio:${repseqVersion}") {
      // exclude logback to prevent excessive logging
      exclude group: "ch.qos.logback", module :"logback-classic"
      exclude group: "ch.qos.logback", module :"logback-core"
   }
   BuildUtils.addExternalDependency(
           project,
           new ExternalDependency(
                   "io.repseq:repseqio:${repseqVersion}",
                   "repseqio",
                   "repseqio",
                   "https://github.com/repseqio/repseqio",
                   ExternalDependency.APACHE_2_LICENSE_NAME,
                   ExternalDependency.APACHE_2_LICENSE_URL,
                   "TCR Analysis"
           )
   )
```

**Comment by labkey-susanh** [comment]
1. Yes, we generally want all the external dependencies recorded and credited on the Credits page, so converting all the declarations for the jars that end up in your module's `lib` directory is called for.
2. the `BuildUtils.addExternalDependency` method can take a third argument, which is a closure for declaring exclusions and the like, so something like this should work:
```
BuildUtils.addExternalDependency(
           project,
           new ExternalDependency(
                   "io.repseq:repseqio:${repseqVersion}",
                   "repseqio",
                   "repseqio",
                   "https://github.com/repseqio/repseqio",
                   ExternalDependency.APACHE_2_LICENSE_NAME,
                   ExternalDependency.APACHE_2_LICENSE_URL,
                   "TCR Analysis"
           ), 
           {
                   // exclude logback to prevent excessive logging
                  exclude group: "ch.qos.logback", module :"logback-classic"
                  exclude group: "ch.qos.logback", module :"logback-core"
         }
   )
```

**Comment by bbimber** [comment]
Thanks @labkey-susanh. I should be able to convert all of these today, and will merge to develop.

One tangential question: within intellij (i am using the LK default, i think), code completion doesnt work in LK gradle files. should it, and/or is there some setting or library I need to link up?

**Comment by bbimber** [comment]
Hi @labkey-susanh 

After updating, I'm getting a weird test failure in org.labkey.core.notification.EmailServiceImpl$TestCase. See here:

https://teamcity.labkey.org/buildConfiguration/LabKey_217Release_External_Discvr_ExternalModulesTestSqlserver/1572405?expandedTest=build%3A%28id%3A1572405%29%2Cid%3A5879

java.lang.LinkageError: loader constraint violation: when resolving method 'void javax.mail.BodyPart.setDataHandler(javax.activation.DataHandler)' the class loader org.labkey.bootstrap.LabKeyBootstrapClassLoader @12299890 of the current class, org/labkey/core/notification/EmailServiceImpl$EmailMessageImpl, and the class loader java.net.URLClassLoader @2077d4de for the method's defining class, javax/mail/Part, have different Class objects for the type javax/activation/DataHandler used in the signature (org.labkey.core.notification.EmailServiceImpl$EmailMessageImpl is in unnamed module of loader org.labkey.bootstrap.LabKeyBootstrapClassLoader @12299890, parent loader java.net.URLClassLoader @2077d4de; javax.mail.Part is in unnamed module of loader java.net.URLClassLoader @2077d4de, parent loader 'app')
  at org.labkey.core.notification.EmailServiceImpl$EmailMessageImpl.createMessage(EmailServiceImpl.java:312)

I am suspicious of this:

New version:
https://github.com/BimberLab/DiscvrLabKeyModules/blob/discvr-21.7/omerointegration/build.gradle

Old version:
https://github.com/BimberLab/DiscvrLabKeyModules/blob/develop/omerointegration/build.gradle

Does this seem like it should be valid?


```
    BuildUtils.addExternalDependency(
            project,
            new ExternalDependency(
                    "org.openmicroscopy:omero-gateway:5.6.3",
                    "OMERO Client",
                    "OMERO Client",
                    "https://docs.openmicroscopy.org/omero/5.6.3/developers/Java.html",
                    ExternalDependency.GNU_LESSER_GPL_21_NAME,
                    ExternalDependency.GNU_LESSER_GPL_21_URL,
                    "Used to interact with OMERO Servers"
            ),
            {
                exclude group: 'OME'
                exclude group: 'antlr'
                exclude group: 'asm'
                exclude group: 'backport-util-concurrent'
                exclude group: 'batik'
                exclude group: 'cglib'
                exclude group: 'ch.qos.logback', module: 'logback-classic'
                // added to prevent excessively verbose global logging
                exclude group: 'com.codahale.metrics'
                exclude group: 'com.drewnoakes'
                exclude group: 'com.esotericsoftware.kryo'
                exclude group: 'com.google.guava'
                exclude group: 'com.jamonapi'
                exclude group: 'com.mortennobel'
                exclude group: 'com.sun.activation', module: 'javax.activation'
                exclude group: 'com.zeroc', module: 'freeze'
                exclude group: 'com.zeroc', module: 'icefreeze'
                exclude group: 'com.zeroc', module: 'icegrid'
                exclude group: 'com.zeroc', module: 'icestorm'
                exclude group: 'commons-beanutils'
                exclude group: 'commons-codec'
                exclude group: 'commons-collections'
                exclude group: 'commons-io'
                exclude group: 'commons-pool'
                exclude group: 'dom4j'
                exclude group: 'edu.ucar'
                exclude group: 'freemarker'
                exclude group: 'geronimo-spec'
                exclude group: 'gnu.getopt'
                exclude group: 'javassist'
                exclude group: 'javax.jts'
                exclude group: 'joda-time'
                exclude group: 'net.sf.ehcache'
                exclude group: 'ome', module: 'formats-gpl'
                exclude group: 'ome', module: 'jai_imageio'
                exclude group: 'ome', module: 'lwf-stubs'
                exclude group: 'ome', module: 'turbojpeg'
                exclude group: 'omero', module: 'dsl'
                exclude group: 'omero', module: 'omero-shares'
                exclude group: 'org.apache.lucene'
                exclude group: 'org.apache.httpcomponents'
                exclude group: 'org.codehaus.btm'
                exclude group: 'org.hibernate'
                exclude group: 'org.hibernate.javax.persistence'
                exclude group: 'org.ini4j'
                exclude group: "org.json", module: "json"
                exclude group: 'org.postgresql'
                exclude group: 'org.scijava'
                exclude group: 'org.subethamail'
                exclude group: 'pdfbox'
                exclude group: 'quartz'
                exclude group: 'xerces'
                exclude group: 'xalan'
                exclude group: 'xml-apis'
                exclude group: 'zeroc', module: 'ice-db'

                exclude group: 'org.springframework', module: 'spring-aop'
                exclude group: 'org.springframework', module: 'spring-context'
                exclude group: 'org.springframework', module: 'spring-expression'
                exclude group: 'org.springframework', module: 'spring-core'
                exclude group: 'org.springframework', module: 'spring-beans'
                exclude group: 'org.springframework', module: 'spring-jdbc'
                exclude group: 'org.springframework', module: 'spring-tx'
            }
    )
```

**Comment by bbimber** [comment]
Note, the diff on my full gradle changes are below if it helps:

https://github.com/BimberLab/DiscvrLabKeyModules/commit/94b6f6e5305a802bdac8eae1515718cab8a711c1

https://github.com/bimberlabinternal/BimberLabKeyModules/commit/3e4b6b542731212ba936c8d5ed1d7bcccab944ec

**Comment by labkey-susanh** [comment]
> Thanks @labkey-susanh. I should be able to convert all of these today, and will merge to develop.
> 
> One tangential question: within intellij (i am using the LK default, i think), code completion doesnt work in LK gradle files. should it, and/or is there some setting or library I need to link up?

In the ideal world, yes, it should, but using the Groovy DSL for Gradle I've not found a way to make this work.

**Comment by labkey-susanh** [comment]
As for the classloader problem, I'm not seeing what the problem is there, but for troubleshooting, I would probably first run the dependencies command in the modules you are changing to see if the dependencies look like you expect them to.  Perhaps `javax.activation` is not getting excluded as expected or there's a different version being brought in somewhere instead of the version 1.2.1 that the api module brings in transitively.

**Comment by bbimber** [comment]
Thanks for following up. I should have posted back here, but I found the problem yesterday. It was one of the other modules where I didn't convert the exclude properly. I'll merge this to develop later today once my lab wakes up and someone approves the prs

**Pull request merged**
